core/iwasm/aot/aot_runtime.c

```diff
@@ -3751,12 +3845,50 @@ aot_free_frame_internal(WASMExecEnv *exec_env)
     exec_env->cur_frame = (struct WASMInterpFrame *)prev_frame;
 }
 
+static inline void
+aot_free_tiny_frame(WASMExecEnv *exec_env)
+{
+    exec_env->wasm_stack.top =
+        get_prev_frame(exec_env, exec_env->wasm_stack.top);
+}
+
 void
 aot_free_frame(WASMExecEnv *exec_env)
 {
-    aot_free_frame_internal(exec_env);
+    if (is_tiny_frame(exec_env)) {
+        aot_free_tiny_frame(exec_env);
+    }
+    else {
+        aot_free_standard_frame(exec_env);
+    }
 }
 
+/**
+ * Releases a frame from the top of the stack only if the top frame
+ * describes imported function.
+ *
+ * This function is used for a frame-per-function in the call_indirect
+ * implementation; it's not possible to know at compile time if the the
+ * opcode calls to imported function (where the frame must be released,
+ * because frame allocation is not part of the function) or the function
+ * within the module (in that case the function itself is responsible for
+ * allocating / freeing the frame), so we need to check it at runtime.
+ */
```

Comment:
If the frame of the import function isn't allocated, there may be several issues: (1) the dumped call stacks may be not complete, (2) for GC, the native function's parameters are not committed frame and AOT runtime can not traversing them and add them to GC's root set, GC reclaim process may be invalid, (3) for performance profiling, the execution time of the import function isn't found.
Can we allocate the import function's frame before calling the function and freeing after calling the function? Like what current code does for both import and non-import functions.

Reply:
> Can we allocate the import function's frame before calling the function and freeing after calling the function? Like what current code does for both import and non-import functions.

@wenyongh that's exactly what happens now; for `call` opcode we have:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/eb4d5bbc8aedef23bf2864c7923ddbc8300004de/core/iwasm/compilation/aot_emit_function.c#L1485-L1503

so the frame will be allocated using `aot_alloc_frame_per_function_frame_for_aot_func` if this is imported function. For `call_indirect` there's *always* a call to `aot_alloc_frame` (there's no condition that checks if it's `frame-per-function` mode), and in the `aot_alloc_frame` the frame allocation is only skipped for non-imported function; otherwise it's allocated:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/eb4d5bbc8aedef23bf2864c7923ddbc8300004de/core/iwasm/aot/aot_runtime.c#L3803-L3822

Reply:
Yes, but I think we can add check in the compiled LLVM IR of opcode call_indirect: compare the function index with import_func_count, and call aot_alloc_frame if func_idx < import_func_count. It doesn't necessary always calling aot_alloc_frame.

And seems no need to add the flag `comp_ctx->call_stack_features.frame_per_function`, my idea is that: (1) in the new code of aot compiler, for both full frame and tiny frame, (a) always allocate frame in the beginning of a non-imported function, (b) always allocate frame before calling the imported function (in both op call and op call_indirect), (c) always emit the WASM_FEATURE_FRAME_PER_FUNCTION flag in the aot file. This simplify the aot compilation and seems no need to aot_alloc_import_frame.

(2) in aot runtime, (a) always aot_alloc_frame before calling import function, (b) aot_alloc_frame before calling non-import function if WASM_FEATURE_FRAME_PER_FUNCTION flag isn't detected, so as to keep backward compatibility.

Reply:
I think it should be simpler, please correct me if I am wrong or there is something I missed.

Reply:
> Yes, but I think we can add check in the compiled LLVM IR of opcode call_indirect: compare the function index with import_func_count, and call aot_alloc_frame if func_idx < import_func_count. It doesn't necessary always calling aot_alloc_frame.

That's I think possible, but perhaps will add slightly more code. I don't mind changing it though as in our usecase `call_indirect` are quite rare.

> And seems no need to add the flag comp_ctx->call_stack_features.frame_per_function

Yes, it could potentially be removed; however, I didn't do it (just yet) due to multimodule / dynamic linking - if we always use frame-per-function mode, the callstack will contain duplicates. So I think we should keep both modes in the compiler for now until we figure out how to do it correctly; there are few options that come to my mind:
* explicitly expose frame-per-function as parameter to wamrc, so users can choose the mode
* allow users to define which functions are provided by another module and which functions are imported from native or modules without instrumentation

Either is fine for me, but also I don't think any of the option needs to be implemented right now and can be added in the future without breaking the ABI. What do you think?

Reply:
Yes, for me the first option (expose frame-per-function as parameter to wamrc) is better and can be added in the future. The latter may be a little complex for developer to choose which functions?

Reply:
agree, I'll make it a follow-up task for the https://github.com/bytecodealliance/wasm-micro-runtime/issues/3758

Reply:
@wenyongh I also removed `aot_free_import_frame` function and moved the allocation of the frame for imported function to the AOT code. Let me know if you have any questions.